### ISSUE TITLE
Should "root types exist" be a formal validation?

### ISSUE THREAD

**Issue by rmosolgo** [facebook/graphql#218]
A GraphQL schema _must_ have a query root, but mutation and subscription roots are optional. If a schema doesn't have a mutation root, `mutation` operations are invalid. Same for `subscription`: if there's no `subscription` root type, we should reject a document with an operation on `subscription`.

We have this validation in graphql-ruby _now_ (thanks @cjoudrey) but we didn't at first, because I was _literally_ reading the spec and implementing the rules one-by-one. 

Should that rule be in the spec? If so, I'm happy to try writing up (I think I have a bow tie around here somewhere 😆 ).

**Comment by stubailo** [comment]
As a side note I think it's reasonable to have a server with only mutation or subscription types, I'm not sure why Query is singled out as being required.

**Comment by erydo** [comment]
Agreed. We've been bitten by this when exposing mutation-only schemas for backend services. The workaround is to create an empty query type, but I'm not sure what the rationale is for that to be explicitly required unlike `mutationType` or `subscriptionType`. Does `queryType` have any other special characteristics?

To be clear, the issue at hand is the non-nullability of the `queryType` field in the Introspection schema:
  - [Section 4 -- Introspection.md](https://github.com/facebook/graphql/blob/e1dd45286e83cea73b7b071dc14e9f4dc5208ace/spec/Section%204%20--%20Introspection.md#schema-introspection)
 - [graphql-js: introspection.js](https://github.com/graphql/graphql-js/blob/8122ba7f6fddbb2b0de9653b9d935cb2aec77be2/src/type/introspection.js#L47)
  - [graphql-python: introspection.py](https://github.com/graphql-python/graphql-core/blob/dadb918764fc95b33d31577e11fd08323b59b081/graphql/type/introspection.py#L41)
